### PR TITLE
add ignored chars to the AST

### PR DESCRIPTION
**Pull request by andimarek**
This adds all the ignored chars to the Ast. Previously we would just skip them, now they are available on the AST.
This is allows for printing out the original Document (not implemented here) and also allows some analysing of the complete Document.

### CODE REVIEW

**Comment by andimarek** [comment]
@bbakerman please have a look. It is ready for merge except the build file modifications.

File: build.gradle [hunk 1]
```diff
@@ -20,7 +20,7 @@ def getDevelopmentVersion() {
         println "git hash is empty: error: ${error.toString()}"
         throw new IllegalStateException("git hash could not be determined")
     }
-    new SimpleDateFormat('yyyy-MM-dd\'T\'HH-mm-ss').format(new Date()) + "-" + gitHash
+    new SimpleDateFormat('yyyy-MM-dd\'T\'HH-mm-ss').format(new Date()) + "-" + gitHash + "-ignored-chars"
```
Comment (bbakerman) [inline build.gradle]: This will need to be removed at some stage right?

File: src/main/antlr/GraphqlCommon.g4 [hunk 2]
```diff
@@ -120,14 +120,17 @@ fragment SourceCharacter :[\u0009\u000A\u000D\u0020-\uFFFF];
 
 Comment: '#' ~[\n\r\u2028\u2029]* -> channel(2);
 
-Ignored: (UnicodeBOM|Whitespace|LineTerminator|Comma) -> skip;
+//Ignored: (UnicodeBOM|Whitespace|LineTerminator|Comma) -> channel(3);
```
Comment (bbakerman) [inline src/main/antlr/GraphqlCommon.g4]: whats with the `//` bit - is it in or out?

File: src/main/java/graphql/language/AbstractNode.java [hunk 3]
```diff
@@ -13,11 +13,13 @@
...
     private final SourceLocation sourceLocation;
     private final List<Comment> comments;
+    private final IgnoredChars ignoredChars;
 
-    public AbstractNode(SourceLocation sourceLocation, List<Comment> comments) {
+    public AbstractNode(SourceLocation sourceLocation, List<Comment> comments, IgnoredChars ignoredChars) {
         this.sourceLocation = sourceLocation;
         Assert.assertNotNull(comments, "comments can't be null");
         this.comments = new ArrayList<>(comments);
+        this.ignoredChars = Assert.assertNotNull(ignoredChars, "ignoredChars can't be null");
```
Comment (bbakerman) [inline src/main/java/graphql/language/AbstractNode.java]: personally I dont think these messages on asserts add value.  If it DID blow up - it was would obvious that this line is offending.  There is no insight to be ahd I think.

Anyway not a huge problem but I would just do `assertNotNull(someField)`

File: src/main/java/graphql/language/IgnoredChars.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+import graphql.PublicApi;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@PublicApi
+public class IgnoredChars implements Serializable {
```
Comment (bbakerman) [inline src/main/java/graphql/language/IgnoredChars.java]: public API??

I guess it is by proxy

File: src/main/java/graphql/language/TypeName.java [hunk 5]
```diff
@@ -11,13 +11,13 @@
 import java.util.function.Consumer;
 
 @PublicApi
-public class TypeName extends AbstractNode<TypeName> implements Type<TypeName> {
+public class TypeName extends AbstractNode<TypeName> implements Type<TypeName>, NamedNode<TypeName> {
```
Comment (bbakerman) [inline src/main/java/graphql/language/TypeName.java]: nice catch

**Pull request merged**